README.md

```diff
@@ -1,189 +1,180 @@
 ```
 
-### `@set`
+### Decorators adding metadata
 
-The `@set` decorator makes a class field declaration behave as a setting a property when it's called, rather than as `Object.defineProperty`. In particular, setters will be called with a normal property set, whereas `Object.defineProperty` just clobbers setters without calling them. For example:
+Decorators can add metadata about class elements by adding a `metadata` property of the context object that is passed in to them. All of the metadata objects are `Object.assign`'ed together and placed in a property reachable from `[Symbol.metadata]` on the class. For example:
 
 ```mjs
-class SuperClass {
-  set x(value) { console.log(value); }
+@annotate({x: "y"}) @annotate({v: "w"}) class C {
+  @annotate({a: "b"}) method() { }
+  @annotate({c: "d"}) field;
 }
 
-class SubClassA extends SuperClass {
-  x = 1;
-}
+C[Symbol.metadata].class.x                     // "y"
+C[Symbol.metadata].class.v                     // "w"
+C[Symbol.metadata].prototype.methods.method.a  // "b"
+C[Symbol.metadata].instance.fields.field.c     // "d"
+```
 
-class SubClassB extends SuperClass {
-  @set x = 1;
-}
+**NOTE:** The exact format of the annotations object is not very well-thought-out and could use more refinement. The main thing I'd like to illustrate here is, it's just an object, with no particular support library to read or write it, and it's automatically created by the system.
 
-const a = new SubClassA();  // does not log anything
-a.x;                        // 1
+This decorator `@annotate` could be implemented as follows:
 
-const b = new SubClassB();  // logs 1
-b.x;                        // undefined
+```js
+function annotate(metadata) {
+  return (_, context) => {
+    context.metadata = metadata;
+    return _;
+  }
+}
 ```
 
-The `@set` decorator is implemented with `@initialize`, which can decorate public fields. `@initialize` takes a callback as an argument, which is called after the field initializer is evaluated,
+Each time a decorator is called, it is passed a new context object, and after each decorator returns, the context object's `metadata` property is read, and if it's not undefined, it's included in the `[Symbol.metadata]` for that class element.
 
-```mjs
-// set.mjs
+<!--
+The desugarings in this article usually omit it, but metadata is desugared as such:
+class C {
+  method() {  }
+}
+let context = { kind: "method", name: "method", isStatic: false };
+C.prototype.method = annotate(C.prototype.method, context);
+C[Symbol.metadata].prototype.methods.method = {...{a: "b"}, ...(context[Symbol.metadata] ?? {})};
+-->
 
-export decorator @set { @initialize(function(value, key) { this[key] = value }) }
-```
+Note that, since metadata is held on the class, not on the method, the metadata is not visible to earlier decorators. Metadata on classes is added to the constructor after all class decorators have run so that they are not lost by later wrapping.
```

Comment:
Not sure I understand correctly the order of execution and ability of class decorators to read the metadata. So, e.g. I have two following sets of decorators:

#### registrator-decorator.js
```js
const register = Symbol('register');

function listen(f, context) {
  assert(context.kind, 'method');
  console.log(1);

  let callback;

  context.metadata = {
    [register](klass) {
       if (isProvider(klass)) {
          callback = () => console.log('provider);
       } else {
          callback = () => console.log('consumer');
       }
    }
  }

  function wrapper(...args) {
    callback();
    return f.apply(this, args)
  }

  return wrapper;
}

function registrator(klass, context) {
  assert(context.kind, 'class');
  console.log(2);

  const {methods} = klass[Symbol.metadata].prototype;

  for (const method in klass[Symbol.metadata].prototype.methods) {
    if (register in methods[method]) {
      methods[method][register](klass);
    }
  }

  return klass;
}
```
#### form-decorator.js
```js
const option = Symbol('option');

function option(name) {
  return (f, context) => {
    assert(context.kind, 'method');
    console.log(3);
  
    context.metadata = {
      [option]: name,
    }

    return f;
  }
}

function form(klass, context) {
  console.log(4);
  const {methods} = klass[Symbol.metadata].prototype;

  for (const method in klass[Symbol.metadata].prototype.methods) {
    if (option in methods[method]) {
      registerOption(klass, methods[method][option], klass.prototype[method]); 
    }
  }

  return klass;
}
```
#### main.js
```js
import {listen, registrator} from './registrator-decorator.js';
import {option, form}  from './form-decorator.js';

@form
@registrator
class Form extends HTMLElement {
  @listen method1() {}

  @listen @option method2() {}
}
```
There are the following questions I have:

1. Is that approach possible?
2. Can the `@form` and `@registrator` get access to a `klass[Symbol.metadata]` as I wrote?
3. What will the execution order be? I see it as something like:
  ```
  1 // @listen on method1
  3 // @option on method2
  1 // @listen on method2
  2 // @registrator on Form
  4 // @form on Form
  ```

Reply:
I think this deserves more thought and iteration. In general, the details of the metadata are sort of the weakest part of this proposal, but I'm confident that we'll be able to iterate and find a solution.